# [*Causal Inference in Statistics: A Primer*](https://www.wiley.com/en-us/Causal+Inference+in+Statistics%3A+A+Primer-p-9781119186847)

# Judea Pearl, Madelyn Glymour, Nicholas P. Jewell 

# Section 1.3: Probability and statistics

---

![](https://media.wiley.com/product_data/coverImage300/46/11191868/1119186846.jpg)

In [1]:
import math
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib as mpl
import matplotlib.pyplot as plt
plt.style.use('seaborn-darkgrid')

## Study questions

### 1.3.2

| Gender | Highest education achieved | Occurrence |
|--------|----------------------------|------------|
| Male   | Never finished high school |        112 |
| Male   | High school                |        231 |
| Male   | College                    |        595 |
| Male   | Graduate school            |        242 |
| Female | Never finished high school |        136 |
| Female | High school                |        189 |
| Female | College                    |        763 |
| Female | Graduate school            |        172 |
| TOTAL  | -                          |       2440 |

The probabilities are

\begin{align*}
  P(\text{High School}) &= (231 + 189)/2440 = 0.17 \\
  P(\text{High School OR Female}) &= (231 + 136 + 189 + 763 + 172)/2440 = 0.61 \\
  P(\text{High School | Female}) &= 189/(136 + 189 + 763 + 172) = 0.15 \\
  P(\text{Female | High School}) &= 189/(189 + 231) = 0.45
\end{align*}

## Independence and marginal independence

![](01-independence.jpg)

![](02-marginal-independence.jpg)

+ Consider the causal chain FIRE $\to$ SMOKE $\to$ DETECTOR
  ON

  SMOKE is a *mediator*. It makes FIRE and DETECTOR ON
  conditionally independent, i.e.
  \begin{align*}
  P(\text{DETECTOR ON } | \text{ FIRE, SMOKE}) &= P(\text{DETECTOR ON } | \text{ SMOKE})
  \end{align*}

  Is this the *definition* of mediator?

## The law of total probability and other rules

+ If $A$ and $B$ are mutually exclusive, then 

  $$P(A \lor B) = P(A) + P(B)$$

+ It follows that

  $$P(A) = P(A \land B) + P(A \land \neg B)$$

+ More generally, if $B_1, B_2, \ldots, B_n$ is a partition,
  then

  $$P(A) = P(A \land B_1) + P(A \land B_2) + \cdots + P(A
  \land B_n)$$

+ Multiplication rule (1.7):

  $$P(A \land B) = P(A|B)P(B) = P(B|A)P(A)$$

+ Division rule:

  $$P(A|B) = \frac{P(A \land B)}{P(B)}$$

  This is equivalent to viewing conditioning as a *filtering*
  operation. Taking probabilities as frequencies, we are
  taking as the denominator the number of instances where
  $B$ is true (instead of the total number of instances).

+ Characterizing independence: $A$ and $B$ are independent iff

  $$P(A \land B) = P(A)P(B)$$

+ Bayes' rule:

  $$P(A | B) = \frac{P(B|A)P(A)}{P(B)}$$

+ Decomposing: "law of alternatives", or "extending the
  conversation". In the book, *conditionalizing on $B$*:

  $$P(A) = P(A|B_1)P(B_1) + P(A|B_2)P(B_2) + \cdots + P(A|B_n)P(B_n)$$

  for $B_1, B_2, \ldots, B_n$ a partition.

+ See examples: defective gadgets and rolling 2 dice.

  For the example of rolling 2 dice, the 2 rolls are
  independent, so we can just multiply the probabilities:

  ![](03-roll-2-dice.jpg)
  

## Using Bayes' rule

+ Hypothesis and evidence:

  ![](04-bayes.jpg)

## Study questions

### 1.3.3 (a)

  ![](05-study-1.3.3.a.jpg) 

  ![](06-study-1.3.3.a-cont.jpg)

### 1.3.3 (b)

  ![](07-study-1.3.3.b.jpg)

### 1.3.4 (a)

  - 3 cards.

  - Card 1: 2 black faces.

  - Card 2: 2 white faces.

  - Card 3: 1 white face, 1 black face.

  - Select card at random, top face is black.

  - It is either card 1 or card 3.

  - So, bottom face has probability 1/2 of being black.

  - **But, then again**, if the top face is black, there is a
    greater probability that it is card 1, and therefore,
    there might be a greater probability that the other face
    is also black.

  - Hmmmm.

### 1.3.4 (b, c)

  - Using Bayes, calculate probability that face-down side
    is black, given that the face-up side is black:

    $$
    P(C_D = \text{b} \mid C_U = \text{b}) = 
    \frac
    {P(C_U = \text{b} \mid C_D = \text{b})\;P(C_D =
    \text{b})}
    {P(C_U = \text{b})}
    $$
      
    ![](08-study-1.3.4.b.c.jpg)

  - Ha! The following table shows that the answer really is
    $\frac{2}{3}$:

    ![](09-study-1.3.4.b.c-cont.jpg) 

    This was a great example of how intuition may mislead
    us! The answer is as surprising as in the Monty Hall
    problem.

### 1.3.5

  ![](10-study-1.3.5.jpg) 

  ![](11-study-1.3.5-cont.jpg)

## Expected values

+ Expected value of *numerical* random variable $X$ is 
  $$E(X) = \sum_x x \cdot P(X=x)$$

+ Expected value of *numerical* function $g(x)$ of
  *numerical* random variable $X$ is 
  $$E[g(X)] = \sum_x g(x) \cdot P(X=x)$$

+ Expected value of $Y$ conditioned on $X$:
  $$
  E(Y \mid X = x) = \sum_y y \cdot P(Y=y \mid X=x)
  $$

+ Expected value $g = E(X)$ minimizes expected square error
  $E\left[(g - X)^2\right]$.

+ Similarly, $g = E(Y \mid X=x)$ minimizes $E[(g - Y)^2 \mid
  X=x]$.

+ The *median* minimizes expected absolute error $E[|g-X|]$.

## Variance and covariance

+ Given *numerical* random variable $X$ with mean $\mu = E[X]$, then
  $\text{Var}(X) = \sigma^2 = E[(X - \mu)^2] = E[X^2] - E[X]^2$.

+ Standard deviation is $\sigma = \sqrt{\sigma^2}$.

+ Given two random variables $X$ and $Y$, the *covariance*
  is
  $$
  \sigma_{XY} = E[(X - E(X))\cdot(Y - E(Y))]
  $$
  This measures how much $X$ and $Y$ *linearly* covary.
  
+ This can also be written
    $$
    \sigma_{XY} = E[XY] - E[X]E[Y]
    $$

+ From Deveaux, some properties of the covariance:

    ![](26-covariance.png)

+ The *correlation coefficient* is a normalization of the
  above wrt the product of stardard deviations:
  $$
  \rho_{XY} = \frac{\sigma_{XY}}{\sigma_X \sigma_Y}
  $$

+ $\rho_{XY} \in [-1, 1]$.

+ When $X$ and $Y$ are independent, $\rho_{XY} = 0$.

## Study questions

### 1.3.6 (a)

![](12-study-1.3.6.jpg)

![](13-study-1.3.6-cont.jpg)

### 1.3.6 (b)

See solution manual

### 1.3.7

![](14-study-1.3.7.jpg)

![](15-study-1.3.7-cont.jpg)

### 1.3.8 (a)

$E(X) = 3.5$

$E(Y) = E(X) + E(Z) = 7$

#### Pandas implementation of craps

In [2]:
def craps_df(): 
    d = {
        'X': np.array([[d1] * 6 for d1 in range(1,7)]).flatten(),
        'Z': list(range(1, 7)) * 6
    }
    craps = pd.DataFrame(d)
    craps['Y'] = craps['X'] + craps['Z']
    return craps

In [3]:
craps = craps_df()

#### $E[Y \mid X = x]$

For $X=1$ and $Y=X+1$, we have

In [4]:
X = 1
# Get only results where X == x
temp = craps[craps['X'] == X]

# P(X = x)
px1 = len(temp)/len(craps)
print(f'P(X={X}) = {len(temp)}/{len(craps)} = {px1}')

# P(X = x & Y = x + 1)
px1y1 = len(temp[(temp['X'] == X) & (temp['Z'] == X + 1)])/len(craps)
print(f'P(Y={X+1} AND X={X}) = {len(temp[(temp["X"] == X) & (temp["Z"] == X + 1)])}/{len(craps)} = {px1y1}')

P(X=1) = 6/36 = 0.16666666666666666
P(Y=2 AND X=1) = 1/36 = 0.027777777777777776


Now, to compute $E[Y \mid X = x]$ for general $x$, we do
\begin{align*}
E[Y \mid X = x] 
    &= (x + 1) \cdot \frac{1}{6} + \cdots + (x + 6) \cdot \frac{1}{6} \\
    &= \frac{1}{6}(6x + 21) \\
    &= x + \frac{7}{2} \\
    &= x + 3.5
\end{align*}

So, according to this calculation, we have

In [5]:
print('x\t E[Y | X = x]')
for x in range(1, 7):
    print(f"{x}\t {x+7/2}")

x	 E[Y | X = x]
1	 4.5
2	 5.5
3	 6.5
4	 7.5
5	 8.5
6	 9.5


Checking with Pandas:

In [6]:
print('x\t E[Y | X = x]')
for x in range(1, 7):
    temp = craps[craps['X'] == x]
    e = temp['Y'].mean()
    print(f"{x}\t {e}")

x	 E[Y | X = x]
1	 4.5
2	 5.5
3	 6.5
4	 7.5
5	 8.5
6	 9.5


So, for each value $x$ of $X$, we have $E[Y \mid X = x] = x + 3.5$

#### $E[X \mid Y = y]$

![](16-study-1.3.8.jpg)

Checking with Pandas:

In [7]:
craps = craps_df()
print(f"n\t E[X | Y = n]")
for n in range(2, 13):
    temp = craps[craps.Y == n]
    print(f"{n}\t {temp.X.mean()}")

n	 E[X | Y = n]
2	 1.0
3	 1.5
4	 2.0
5	 2.5
6	 3.0
7	 3.5
8	 4.0
9	 4.5
10	 5.0
11	 5.5
12	 6.0


So, for each value $y$ of $Y$, we have $E[X \mid Y = y] = y/2$

#### $\text{Var}(X)$

Remember $\text{Var}(X) \quad=\quad E[(X - E(X))^2] \quad=\quad E[X^2] - E(X)^2$. 

The derivation is

![](17-study-1.3.8-cont.jpg)

So we have
\begin{align*}
    \text{Var}(X) 
        &= E[X^2] - E[X]^2 \\
        &= \frac{1}{6} (1^2 + 2^2 + \cdots + 6^2) - 3.5^2 \\
        &= \frac{1}{6} \cdot 91 - 3.5^2 \\
        &= 2.92
\end{align*}
which checks:

In [8]:
craps = craps_df()
print(f"Var(X) = {craps.X.var(ddof=0):.3}")

Var(X) = 2.92


#### $\text{Var}(Y)$

+ $\text{Var}(Y) = E[Y^2] - E[Y]^2$

+ Let's calculate $E[Y]$ first:
  $$
  \begin{align*}
      E[Y] \quad=\quad 1/36 \quad\cdot\quad (
          & 2 + 3 + 4 + 5 + 6 + 7 + {}\\
          & \phantom{2 + {}} 3 + 4 + 5 + 6 + 7 + 8 + {}\\
          & \phantom{2 + 3 + {}} 4 + 5 + 6 + 7 + 8 + 9 + {}\\
          & \phantom{2 + 3 + 4 + {}} 5 + 6 + 7 + 8 + 9 + 10 + {}\\
          & \phantom{2 + 3 + 4 + 5 + {}} 6 + 7 + 8 + 9 + 10 + 11 + {}\\
          & \phantom{2 + 3 + 4 + 5 + 6 + {}} 7 + 8 + 9 + 10 + 11 + 12)
  \end{align*}
  $$
  
  But that sum is just `craps.Y.sum()`, which equals {{craps.Y.sum()}}, so 
  $$
  E[Y] = \frac{252}{36} = 7.0
  $$
  
  But that should have been obvious, because 
  $$
  E[Y] = E[X + Z] = E[X] + E[Z]
  $$
  
  Plus, I already had this result from [above](#1.3.8-(a)).
  
+ Let's calculate $E[Y^2]$ now:
  $$
  E[Y^2] \quad=\quad \sum_y y^2 P(y) \quad=\quad \frac{1}{36}(2^2 + 2\cdot 3^2 + 3\cdot 4^2 + \cdots + 2\cdot 11^2 + 12^2)
  $$
  Let's compute this in Python, using the `craps` dataframe: 

In [9]:
y_squared = craps.Y**2
print(f"E[Y^2] = {y_squared.sum() / 36 : .3}")

E[Y^2] =  54.8


+ So $\text{Var}(Y) = E[Y^2] - E[Y]^2 = 54.8 - 49 = 5.8$.

+ Let's check it:

In [10]:
print(f"Var(Y) = {craps.Y.var(ddof=0) : .2}")

Var(Y) =  5.8


#### $\text{Cov}(X, Y) = \sigma_{XY}$

+ Remember $\sigma_{XY} = E[(X - E[X])(Y - E[Y])]$

+ But this is equivalent to $\sigma_{XY} = E[XY] - E[X]E[Y]$

+ Well, $E[X]E[Y] = 3.5 \cdot 7 = 24.5$

+ Let's compute $E[XY]$ in Python:

In [11]:
xy = craps.X * craps.Y
print(f"E[XY] = {xy.sum() / 36 : .3}")

E[XY] =  27.4


+ So $\sigma_{XY} = 27.4 - 24.5 = 2.9$.

+ Let's check it in Pandas, but remember that the `cov()` method normalizes by $N - 1$, which gives a different value from ours. So we also present the value normalized by $N$.

In [12]:
cov = craps.X.cov(craps.Y)
print(f"Cov(X, Y) = {cov : .2} (normalized by N-1)")
print(f"Cov(X, Y) = {cov * (len(craps) - 1) / len(craps) : .2} (normalized by N)")

Cov(X, Y) =  3.0 (normalized by N-1)
Cov(X, Y) =  2.9 (normalized by N)


#### $\rho_{XY}$

$$
\rho_{XY} = \frac{\sigma_{XY}}{\sigma_X \sigma_Y} = \frac{2.9}{\sqrt{2.92} \sqrt{5.8}}
$$

In [13]:
print(f"𝜌_𝑋𝑌 = {2.9/(math.sqrt(2.92) * math.sqrt(5.8)): .3}")

𝜌_𝑋𝑌 =  0.705


### 1.3.8 (b)

Results of 12 rolls of two fair dice:

In [14]:
# Results taken from the book
x = [6, 3, 4, 6, 6, 5, 1, 3, 6, 3, 5, 4]
z = [3, 4, 6, 2, 4, 3, 5, 5, 5, 5, 3, 5]
d = {
    'X': x,
    'Z': z
}
craps2 = pd.DataFrame(d)
craps2['Y'] = craps2['X'] + craps2['Z']

craps2

X  Z   Y
0   6  3   9
1   3  4   7
2   4  6  10
3   6  2   8
4   6  4  10
5   5  3   8
6   1  5   6
7   3  5   8
8   6  5  11
9   3  5   8
10  5  3   8
11  4  5   9

In [15]:
print("EXPECTED VALUES")
print("===============")
print(f"E[X] = {craps2.X.mean() : .3}")
print(f"E[Z] = {craps2.Z.mean() : .3}")
print(f"E[Y] = {craps2.Y.mean() : .3}")

print()
print("VARIANCES (N)")
print("=============")
print(f"Var[X] = {craps2.X.var(ddof=0) : .4}")
print(f"Var[Z] = {craps2.Z.var(ddof=0) : .4}")
print(f"Var[Y] = {craps2.Y.var(ddof=0) : .4}")

print()
print("VARIANCES (N-1)")
print("===============")
print(f"Var[X] = {craps2.X.var() : .4}")
print(f"Var[Z] = {craps2.Z.var() : .4}")
print(f"Var[Y] = {craps2.Y.var() : .4}")

# print()
# print("CONDITIONAL EXPECTED VALUES")
# print("===========================")
# for i in range(1, 7):
#     temp = craps2[craps2.X == i]
#     print(f"E[Y | X={i}] = {temp.Y.mean() : .2}")
#     
# print()
# 
# for i in range(2, 13):
#     temp = craps2[craps2.Y == i]
#     print(f"E[X | Y={i}] = {temp.X.mean() : .2}")
    
print()
print("COVARIANCE (N)")
print("==============")
covn = craps2.X.cov(craps2.Y) * (len(craps2) - 1) / len(craps2)
print(f"Cov(X, Y) = {covn : .4}")

print()
print("COVARIANCE (N-1)")
print("================")
cov = craps2.X.cov(craps2.Y)
print(f"Cov(X, Y) = {cov : .4}")

print()
stdx = math.sqrt(craps2.X.var(ddof=0))
stdy = math.sqrt(craps2.Y.var(ddof=0))
print("CORRELATION (N)")
print("===============")
rhon = covn / (stdx * stdy)
print(f"Corr(X, Y) = {rhon : .3}")

print()
print("CORRELATION (N-1)")
print("=================")
rho = cov / (stdx * stdy)
print(f"Corr(X, Y) = {rho : .3}")

print()
print("CORRELATION (Pandas)")
print("====================")
print(craps2.corr())

EXPECTED VALUES
E[X] =  4.33
E[Z] =  4.17
E[Y] =  8.5

VARIANCES (N)
Var[X] =  2.389
Var[Z] =  1.306
Var[Y] =  1.75

VARIANCES (N-1)
Var[X] =  2.606
Var[Z] =  1.424
Var[Y] =  1.909

COVARIANCE (N)
Cov(X, Y) =  1.417

COVARIANCE (N-1)
Cov(X, Y) =  1.545

CORRELATION (N)
Corr(X, Y) =  0.693

CORRELATION (N-1)
Corr(X, Y) =  0.756

CORRELATION (Pandas)
          X         Z         Y
X  1.000000 -0.550516  0.692868
Z -0.550516  1.000000  0.220527
Y  0.692868  0.220527  1.000000


Well, the book has some different results:

![](18-study-1.3.8-cont.png)

The differences are

+ For the variances, the book normalizes by $N$, even though this is a sample, not a population.

+ In Pandas, the choice between $N$ and $N-1$ can be made via the `ddof` optional argument (default is $N - 1$).

+ For the covariances, the book normalizes by $N - 1$.

+ In Pandas, normalization is by $N - 1$ only. 

+ For the correlation coefficient (Pearson), the book again normalizes by $N - 1$.

+ In Pandas, normalization is by $N$.

### 1.3.8 (c, d)

+ $E[Y \mid X = 3] = 6.5$ (from [above](#$E[Y-\mid-X-=-x]$))

+ $E[X \mid Y = 4] = 2.0$ (from [above](#$E[X-\mid-Y-=-y]$))

### 1.3.8 (e)

+ $E[X \mid Y = 4 \land Z = 1] = 3$

## Regression

+ If we know the distribution, then the best prediction of $Y$ given $X=x$ is 
    $$
    E[Y \mid X=x]
    $$
    
+ Usually, we do *not* have the distribution, so we do regression using the data directly.

+ The data are
    $$
    (x_1, y_1), \ldots, (x_n, y_n)
    $$

+ Least squares regression line is
    $$
    y' = \alpha + \beta x
    $$
    such that it minimizes the sum of squared residuals:
    $$
    \sum_i (y_i - y'_i)^2
    $$

+ $\beta$ is the slope of the line.

+ $\alpha$ is the intercept.

### Orthogonality

The book derives formulae for $\alpha$ and $\beta$ using the *orthogonality principle*:

![](19-orthogonality.jpg)

**BUT**

![](20-deveaux-slope.jpg)

### Stuck :-(

I really cannot make sense of this derivation using the orthogonality principle.

Something is wrong, and I don't know what it is. Do you, Mr. Jones?

Later, I will derive formulae for the coefficients using Matrix Calculus and gradients instead.

See http://explained.ai/matrix-calculus/index.html.

Now move on.

### Unstuck :-)

#### Study question 1.3.9 (a)

The answer to study question 1.3.9 shows how to prove that the slope of the best fit line $y = a + bx$ is
$$
b = \frac{\sigma_{XY}}{\sigma^2_X}
$$

The proof has no mention of $\varepsilon$ or the orthogonality principle, though.

Here it is:

![](21-bam.jpg)


I've just found the error in my previous attempt to prove this using the orthogonality principle.

Here is the wrong version again:

![](19-orthogonality.jpg)


The first problem here is that in line 5 I multiplied both sides of the equation by $Y$, when I should have multiplied by $X$. 

The second problem is that in line 7 I assumed $E[Y\varepsilon] = 0$, but this is false!

It is false because I am working from $Y = \alpha + \beta X + \varepsilon$, which makes $Y$ *dependent* on $\varepsilon$.

What the orthogonality principle ensures is that $E[X\varepsilon] = 0$, a fact I used --- implicitly --- in [the correct derivation above](#Unstuck-:-\)).

I say "implicitly" because the initial equation is actually
$$
Y = a + bX + \varepsilon
$$
and, later, when I multiply both sides by $X$ --- *not* by $Y$ ---, I get
$$
E[XY] = aE[X] + bE[X^2] + E[X\varepsilon]
$$
and *now* orthogonality makes $E[X\varepsilon]$ vanish and things finally work out.

## Multiple regression

![](22-multiple-regression.jpg)

### Example of planes and slopes

![](23-example.png)

+ Suppose the red plane is the best fit. Its equation is 
    $$
    Z = 2X + 3Y
    $$
    
+ The blue plane is a plane where $Y$ is held constant.

+ The orange line is the intersection of both planes. 

+ The slope of this line -- i.e., the slope of $Z$ on $X$ when $Y$ is held constant -- is the coefficient of $X$, which is 2.

![](24-example.png)

+ Now the green plane is a plane where $X$ is held constant.

+ The orange line is the intersection of both planes. 

+ The slope of this line -- i.e., the slope of $Z$ on $Y$ when $X$ is held constant -- is the coefficient of $Y$, which is 3.

### Study question 1.3.9 (b)

![](25-1.3.9.png)

The question should reference study question 1.3.8, not 1.3.7.

Just use the values in the formulae. See the solution manual. The coefficients will be $1$ or $-1$, according to the case:

+ If the sum $Y$ is held constant, an increase of 1 in $X$ will mean a decrease of 1 in $Z$.

+ etc.